plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -319,14 +351,30 @@ private static ConnectorPageSource createOrcPageSource(
                         readType = column.getType();
                     }
 
+                    List<List<Integer>> fieldIdProjections = fileColumnsByIcebergId.isEmpty() ?
+                            projectionsByName.get(column.getBaseColumnIdentity().getName()) :
+                            projectionsByFieldId.get(column.getId());
+                    ProjectedLayout projectedLayout = IcebergOrcProjectedLayout.createProjectedLayout(orcColumn, fieldIdProjections);
```

Comment:
This is probably wrong but I'd like to address it in a follow up PR under the other issue: https://github.com/trinodb/trino/issues/9843

Based on the comments from Ryan in there, all of the name based code we have here is probably not correct in the first place.

Reply:
thanks. My only concern is that this will cause regression for users who haven't added the mapping field. Should we at least deprecate first? Or do we just say that dereference pushdown can be only used when IDs are present. i.e. projection pushdown flag serves as deprecation flag for name/id too? cc @findepi

Reply:
The current code handling migrated Hive files (without iceberg IDs) is wrong, produces incorrect results and needs to be rewritten anyway. I don't think it's practical to hold on dereference pushdown until that happens. Rather, I would prefer to temporarily disable handling of migrated files. Would that be better?

Reply:
> I don't think it's practical to hold on dereference pushdown until that happens.

yeah, I don't mean to suggest that at all. 

> I would prefer to temporarily disable handling of migrated files.

My concern is that this will break queries for folks in the process of hive->migration, and can prevent Trino upgrade.

I see your point as per #9843, that name-based mapping is wrong w.r.t. Iceberg spec, but I'd expect this to be very common for Trino users in absence of iceberg ids. So IMO it'll be better to keep supporting it until we've  support for the "correct" mapping using `schema.name-mapping.default` - if it's a relatively low overhead (implementing a couple methods).

Reply:
cc @electrum

Reply:
Let me attempt at consolidating the discussion:

* This only concerns Iceberg tables with files that do not have Iceberg IDs. 

* For tables without “schema.name-mapping.default” property:
  * Before this PR: Trino uses simple name-based mapping. This also seems consistent with Iceberg readers. (e.g. https://github.com/apache/iceberg/pull/580)
  * After this PR: dereference queries will fail on these tables. This'd be a regression in the next release. 

* For tables with “schema.name-mapping.default” property:
  * Before this PR: Trino uses simple name-based mapping. This is incorrect.
  * After this PR: dereference queries will fail on these tables. This is better than silently returning incorrect results, however it is not exhaustive.

* To avoid regression in the first case, we may need to add more changes in the same release. If proper handling of “schema.name-mapping.default” takes more time/effort, we may want to at least restore name-based mapping for the case when  “schema.name-mapping.default” is absent.

Reply:
That is a good summary, thanks. I spent some time this afternoon looking at using `schema.name-mapping.default`. It looks pretty straightforward.

Reply:
I put up a change to use the default name mapping here: https://github.com/trinodb/trino/pull/9959

It's based on this PR

Reply:
> For tables without “schema.name-mapping.default” property:

> Before this PR: Trino uses simple name-based mapping. This also seems consistent with Iceberg readers. (e.g. apache/iceberg#580)

As far as consistency with Spark, that behavior changes depends on the reader, Spark using Iceberg v0.12 has different results reading ORC vs Parquet. According to Ryan https://github.com/trinodb/trino/issues/9843#issuecomment-970514001 the values should be `null`

Reply:
thanks for following up on this!